core/src/commonMain/kotlin/com/ustadmobile/core/viewmodel/signup/SignUpViewModel.kt

```diff
@@ -248,8 +270,69 @@ class SignUpViewModel(
         }
     }
 
+    fun onUsernameChanged(newValue: String) {
+        val filteredValue = filterUsernameUseCase(
+            username = newValue,
+            invalidCharReplacement = ""
+        )
+
+        val updatedPerson = _uiState.value.person?.shallowCopy {
+            username = filteredValue
+        }
+        if (_uiState.value.person?.username!=filteredValue){
```

Comment:
Here it would be better to do one _uiState.update call, and put the if statement in that. 

e.g.

_uiState.update { prev -> 
  prev.copy(
     usernameSetByUser = if(...),
     person = ...  
}

Everytime you call update - that will trigger compose to recompose. Better to only do that once.